data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java

```diff
@@ -23,10 +29,11 @@
     private static final String PARAMETERS_KEY = "Parameters";
     public static final String OPTION_SERVER_TIMEOUT = "servertimeout";
     public static final String OPTION_CLIENT_REQUEST_ID = "ClientRequestId";
-    private static final long NANOS_TO_MILLIS = 1000000L;
-    private HashMap<String, Object> parameters;
-    private HashMap<String, Object> options;
-
+    private final HashMap<String, Object> parameters;
+    private final HashMap<String, Object> options;
+    private static final Pattern PATTERN =
+            Pattern.compile("(?:(\\d+)\\.)?((?:[0-2]?\\d:)?(?:60|[0-5]\\d):(?:60|[0-5]\\d)(?:\\.\\d+)?)",
```

Comment:
This regex is a lot, if you took it from somewhere you should link it, and if you made it then it needs comments.
I think it's subtly wrong though:

1. 60 isn't a valid value for minutes/seconds (unless there's a leap second, maybe?)
2. Since the hour is optional, but isn't dependent on the days, values like "1.53:00" are valid even though they seem not to


But I think that we can do it without the regex - you already validate the string with Long.parseLong and 
LocalTime.parse, so you don't need to do the work twice.

I assume you did this because LocalTime.parse doesn't support days, so instead of the regex just split on the first ".".

There is an edge case of when they don't have days, but do have millis (13:24:30.54), but in this case just check if the days have a ":", and if so don't split.

Reply:
I would name this MILLIS_FROM_TIMESPAN_REGEX instead of PATTERN.

In terms of Asaf's comments:
- I think the validity of "60" depends on the specs of the producer (C#'s System.Timespan). It might be correct - I'm sure it's documented somewhere.
- I'm surprised the hour component *can* be optional! And as Asaf said, especially if the days are present.
- I suspect the above are incorrect, per some regexes that others built without allowing "60" or independently-optional hours: https://stackoverflow.com/questions/12438861/timespan-regex-with-days-portion
- I personally would keep the Regex, but I would refactor `getMillisFromTimespanString` to create built-in Java methods, like Temporal and Duration, from the regex' 2 (or you might need more) groups. Getting millis from these standard objects would be trivial.

I'm personally less concerned with accepting an invalid C# Timespan (like hour==29), than with being sure we properly convert all valid Strings accurately to millis.

Reply:
Also, I believe the servertimeout property may not be more than an hour!

If that's the case, we can both validate the property's input, and simplify this regex.

Reply:
The optional indicator (?) is after:
- (?:(\\d+)\\.)?
- (?:[0-2]?\\d:)?
- (?:\\.\\d+)?
Aren't the first 2 the days and hours?

Reply:
I didnt get your point - you can test the regex in some GUI

Reply:
The RegEx is: `(?:(\\d+)\\.)?((?:[0-2]?\\d:)?(?:60|[0-5]\\d):(?:60|[0-5]\\d)(?:\\.\\d+)?)`
Its parts seem to be:
- (?:(\\d+)\\.)? == Days, which are optional
- ((?:[0-2]?\\d:)? == Hours, which are optional
- (?:60|[0-5]\\d): == Minutes
- (?:60|[0-5]\\d) == Seconds
- (?:\\.\\d+)?) == < Seconds, which are optional

Am I wrong?